test/lit/passes/vacuum-tnh.wast

```diff
@@ -278,4 +284,345 @@
+  ;; NO_TNH-NEXT:    (i32.store
+  ;; NO_TNH-NEXT:     (i32.const 0)
+  ;; NO_TNH-NEXT:     (i32.const 1)
+  ;; NO_TNH-NEXT:    )
+  ;; NO_TNH-NEXT:    (i32.store
+  ;; NO_TNH-NEXT:     (i32.const 2)
+  ;; NO_TNH-NEXT:     (i32.const 3)
+  ;; NO_TNH-NEXT:    )
+  ;; NO_TNH-NEXT:    (unreachable)
+  ;; NO_TNH-NEXT:   )
+  ;; NO_TNH-NEXT:  )
+  ;; NO_TNH-NEXT: )
+  (func $block-unreachable-all (param $p i32)
+    (if
+      (local.get $p)
+      (block
+        ;; Both stores can be removed, and even the entire if arm and then the
+        ;; entire if.
+        (i32.store
+          (i32.const 0)
+          (i32.const 1)
+        )
+        (i32.store
+          (i32.const 2)
+          (i32.const 3)
+        )
+        (unreachable)
+      )
+    )
+  )
+
+  ;; YESTNH:      (func $block-unreachable-but-call
+  ;; YESTNH-NEXT:  (i32.store
+  ;; YESTNH-NEXT:   (i32.const 0)
+  ;; YESTNH-NEXT:   (i32.const 1)
+  ;; YESTNH-NEXT:  )
+  ;; YESTNH-NEXT:  (call $block-unreachable-but-call)
+  ;; YESTNH-NEXT:  (unreachable)
+  ;; YESTNH-NEXT: )
+  ;; NO_TNH:      (func $block-unreachable-but-call
+  ;; NO_TNH-NEXT:  (i32.store
+  ;; NO_TNH-NEXT:   (i32.const 0)
+  ;; NO_TNH-NEXT:   (i32.const 1)
+  ;; NO_TNH-NEXT:  )
+  ;; NO_TNH-NEXT:  (call $block-unreachable-but-call)
+  ;; NO_TNH-NEXT:  (i32.store
+  ;; NO_TNH-NEXT:   (i32.const 2)
+  ;; NO_TNH-NEXT:   (i32.const 3)
+  ;; NO_TNH-NEXT:  )
+  ;; NO_TNH-NEXT:  (unreachable)
+  ;; NO_TNH-NEXT: )
+  (func $block-unreachable-but-call
+    ;; A call cannot be removed, even if it leads to a trap, since it might
+    ;; throw or call an import. We can remove the store after it, though.
+    (i32.store
+      (i32.const 0)
+      (i32.const 1)
+    )
+    (call $block-unreachable-but-call)
+    (i32.store
+      (i32.const 2)
+      (i32.const 3)
+    )
+    (unreachable)
+  )
+
+  ;; YESTNH:      (func $catch-pop
+  ;; YESTNH-NEXT:  (try $try
+  ;; YESTNH-NEXT:   (do
+  ;; YESTNH-NEXT:    (call $catch-pop)
+  ;; YESTNH-NEXT:   )
+  ;; YESTNH-NEXT:   (catch $tag
+  ;; YESTNH-NEXT:    (drop
+  ;; YESTNH-NEXT:     (pop i32)
+  ;; YESTNH-NEXT:    )
+  ;; YESTNH-NEXT:    (unreachable)
+  ;; YESTNH-NEXT:   )
+  ;; YESTNH-NEXT:  )
+  ;; YESTNH-NEXT: )
+  ;; NO_TNH:      (func $catch-pop
+  ;; NO_TNH-NEXT:  (try $try
+  ;; NO_TNH-NEXT:   (do
+  ;; NO_TNH-NEXT:    (call $catch-pop)
+  ;; NO_TNH-NEXT:   )
+  ;; NO_TNH-NEXT:   (catch $tag
+  ;; NO_TNH-NEXT:    (drop
+  ;; NO_TNH-NEXT:     (pop i32)
+  ;; NO_TNH-NEXT:    )
+  ;; NO_TNH-NEXT:    (i32.store
+  ;; NO_TNH-NEXT:     (i32.const 0)
+  ;; NO_TNH-NEXT:     (i32.const 1)
+  ;; NO_TNH-NEXT:    )
+  ;; NO_TNH-NEXT:    (unreachable)
+  ;; NO_TNH-NEXT:   )
+  ;; NO_TNH-NEXT:  )
+  ;; NO_TNH-NEXT: )
+  (func $catch-pop
+    (try $try
+      (do
+        ;; Put a call here so the entire try-catch is not removed as trivial.
+        (call $catch-pop)
+      )
+      (catch $tag
+        ;; A pop on the way to a trap cannot be removed. But the store can.
```

Comment:
I thought the spec said that when you catch something it is placed on the stack, so it has to be popped? Or are you saying that the unreachable after it would handle that?

Reply:
The unreachable would handle that.

Reply:
I see. Ok, I'll put a TODO here.